Refactor case study 1 (#116)
ImperialCollegeLondon/digital_research_sustainability_training: episodes/05_case_study1.md

- Celia should consider all of the hardware that she uses to develop her software.
- For the embodied emissions she should consider the source the device (was it newly
  purchased, refurbished, inherited from another group member?), its age, and try to get
  information from the manufacturer's Product Carbon Footprint (PCF) datasheet.
- For the operational emissions, if using local hardware, Celia could use a power meter
  to measure the power used during development. Given the general skew towards embodied
  emissions in consumer devices though its less important to be precise with this value.
  Bearing this in mind she could use a simpler method to estimate such as the [Green
  Algorithms Calculator], or taking a proportion of embodied emissions. In any case it
  would be useful to track how much time she spends doing development and the amount of
  computational work involved. She can use the measured or estimated power usage with a
  local carbon intensity value to get emissions.

### GitHub Actions

- Celia should check where her workflows are running. If using self-hosted runners it
  may be possible to get information about the hardware and utilisation. More likely

>   however in the case of GitHub hosted runners it will be difficult or impossible to

Reviewer: ```suggestion
  however in the case of GitHub hosted runners it will be difficult or impossible to find
```

  such information.
- One data point that is easy to collect is the total runtime of her workflows over a
  given period. Using assumptions about hardware and utilisation she could use this with
  the [Green Algorithms Calculator] but these values would be very approximate and would
  not account for embodied emissions.
- [ECO CI] is a specialist tool that could be added to her workflows to get better
  estimates with less work!

### LLM Use

> - The simplest method would be to [Hugging Face Ecologits calculator]. The key

Reviewer: ```suggestion
- The simplest method would be to use the [Hugging Face Ecologits calculator]. The key
```

  information to track here is which models are used and ideally the number of tokens
  generated during use. In practice, getting token counts is not straightforward so

>   tracking the number of queries and the length of the responses sufficient.

Reviewer: ```suggestion
  tracking the number of queries and the length of the responses would be sufficient.
```

- One alternative might be to make some assumptions about the amount of compute required
  per response and the hardware used and plugging this into the [Green Algorithms
  Calculator].
- Another alternative might be to run a local LLM and measure the power usage but this
  would be a lot of work and of dubious accuracy.
- Note that none of the above methods account for training emissions. In practice there
  is poor data available for this and it's not clear what proportion of the training
  emissions it would make sense to associate with Celia's usage.

### Software Usage

- In order to make comparable estimates for both her local and the remote groups' usage
  Celia should try and use a consistent methodology for both.
- She could measure the energy usage (using a power meter or [codecarbon]) from some

>   local runs use this as a basis for estimating remote runs (assuming they're running on

Reviewer: ```suggestion
  local runs and use this as a basis for estimating remote runs (assuming they're running on
```

  similar hardware).

> - Alternatively as she is likely to be making a lot of assumptions anyway she could use

Reviewer: ```suggestion
- Alternatively, as she is likely to be making a lot of assumptions anyway she could use
```

  the [Green Algorithms Calculator] to get ballpark estimates for local and remote
  usage. She will have to consider carefully what carbon intensity value(s) to use,
  assuming an average may make the most sense.
- In either case she should collect as much information as possible about the hardware
  being used for both local and remote usage as well as information about how often the
  software is run and for how long. It should be relatively easy to get this information
  for local usage and she may be able to get information from some remote users.

> - Similarly for embedded emissions she could assume the local hardware to be

Reviewer: ```suggestion
- Similarly, for embedded emissions she could assume the local hardware to be
```

>   representative for remote usage or use some generic average values. Practically, as

Reviewer: ```suggestion
  representative of the remote usage or use some generic average values. Practically, as
```

[codecarbon]: https://github.com/mlco2/codecarbon

::::::::::::::::::::::::::::::::::::::::

:::::::::::::::::::::::::::::::::::::::::::::::

## Analysis

::::::::::::::::::::::::::::::: challenge

## Estimating emissions (20 minutes)

Using the information and methodolgies provided below produce an estimate for the carbon
emissons for each of Celia's activities.

### Software Development

- Celia exclusively uses a laptop for development - an HP EliteBook 840 G9 - that was
  newly purchased for the project. She estimates that she uses it for around 20 hours
  per week for software development intermixed with other tasks.
- The development process is **not particularly computationally intensive** as she
  mostly works with an Integrated Development Environment and runs the test suite
  occasionally.

**Use the PCF sheet for the laptop and the [Green Algorithms Calculator] to estimate the
emissions associated with software development. You will have to make some assumptions
and decide how to deal with the mixed use of the laptop.**

### Use of GitHub Actions

- Celia's workflows run on any push to a branch, when a pull request is opened and when
  a release is created.

> - Celia's workflows run on GitHub hosted runners.

Reviewer: ```suggestion
- The workflows run on GitHub hosted runners.
```

  utilisation she decides to err on the side of caution and assume her development
  activities use a full CPU core for the full 20 hours she spends developing. This
  provides an estimate of 58 gCO₂e of operational emissions per week.

[PCF datasheet for the HP EliteBook 840 G9]: https://h20195.www2.hp.com/v2/GetDocument.aspx?docname=c09266068

### Emissions from GitHub Actions

- Given she has an estimate for a workflow of 500 seconds she chooses to simply scale
  this up to the full runtime of 1640 seconds. This gives an estimate of around 6 gCO₂e
  per week.

### Emissions from LLM usage

- Using the Hugging Face EcoLogits calculator Celia estimates the emissions from her
  weekly LLMs usage to be around 1 gCO₂e.

### Emissions from the users of the software package

- Celia has enough details to estimate her groups activities using the Green Algorithms
  calculator. Doing this for the known runtime and hardware of her research group this
  provides an estimate of 569.61 gCO₂e per week.
- To estimate the impact of other users of her software she could consider using this
  number as a reference although it might make for a pretty rough estimate. This would

>   give an estimate of` around 17 kgCO₂e per week of operational emissions.

Reviewer: ```suggestion
  give an estimate of around 17 kgCO₂e per week of operational emissions.
```

### Other

Celia also integrates the [codecarbon] as an optional dependency in her code base so

> that it can report the carbon emissions when the code is run. This allows her more

Reviewer: ```suggestion
that it can report the carbon emissions when the code is run. This allows her to more
```

> easily to track the emissions associated with the usage of her package.

Reviewer: ```suggestion
easily track the emissions associated with the usage of her package.
```
